### ISSUE TITLE
Valgrind On-boarding Project for Intel

### ISSUE THREAD

**Issue by jjscheel** [riscv-admin/dev-partners#31]
### Technical Group

Applications & Tools HC

### ratification-pkg

Vector

### Technical Liaison

Haicheng

### Task Category

Other

### Task Sub Category

- [ ] gcc
- [ ] binutils
- [ ] gdb
- [ ] intrinsics
- [ ] Java
- [ ] KVM
- [ ] ld
- [ ] llvm
- [ ] Linux kernel
- [ ] QEMU
- [ ] Spike

### Ratification Target

3Q2023

### Statement of Work (SOW)

**Component names:**
Valgrind

**Requirements:**
 1. Design and implement a general framework to support RVV instructions on Valgrind
including binary disassembly, IR encoding and instrumentation. 
 2. Enable valgrind memcheck and riscv vector unit-stride load/store instructions to demonstrate Valgrind
RVV framework functionality.

**Deliverables:**
  - General framework to support RVV instructions in Valgrind
  - Valgrind memcheck with RVV unit-stride load/store instruction support
  - Support 32 vector registers, and related unprivileged CSRs (vtype, vl, vlenb)
  - Vector Configuration-Setting Instructions Support (vsetvli/vsetivli/vsetvl)
  - Vector unit-stride Load/Store Instructions Support (vle8/vle16/vle32/vle64/vse8/vse16/vse32/vse64/vle8ff/vmseq/vmseq/vmsne/vmor/vfirst/vlm/vsm/vmsif/vmv)
 

**Acceptance Criteria:**
  - Such test cases get pass in [https://github.com/riscv-non-isa/rvv-intrinsic-doc/tree/master/examples](https://github.com/riscv-non-isa/rvv-intrinsic-doc/tree/master/examples):
      - [rvv_memcpy.c](https://github.com/riscv-non-isa/rvv-intrinsic-doc/blob/master/examples/rvv_memcpy.c)
      - [rvv_strcmp.c](https://github.com/riscv-non-isa/rvv-intrinsic-doc/blob/master/examples/rvv_strcmp.c)
      - [rvv_strcpy.c](https://github.com/riscv-non-isa/rvv-intrinsic-doc/blob/master/examples/rvv_strcpy.c)
      - [rvv_strlen.c](https://github.com/riscv-non-isa/rvv-intrinsic-doc/blob/master/examples/rvv_strlen.c)
      - [rvv_strncpy.c](https://github.com/riscv-non-isa/rvv-intrinsic-doc/blob/master/examples/rvv_strncpy.c)
  - Memcheck can detect out the errors injected by some modifications based on above test cases.
 

**Projected timeframe:** 2023-06-30

**SOW Signoffs:**
  - Task group liaison sign-off: 
  - Development partner sign-off
  - Application & Tools HC sign-off (if POC or toolchain/library work): 

### Waiver

- [ ] Freeze
- [ ] Ratification

**Comment by jjscheel** [comment]
@haicheng-li, I see that.  We can do 2 things:
1. I can input your text from the document you sent me
2. You can create your own item which will allow you to edit your comments

Let me know if you want me to do the first item.

**Comment by haicheng-li** [comment]
> * I can input your text from the document you sent me

Pls. help input the text then. Thanks.

**Comment by jjscheel** [comment]
@ptomsich, can you review the contents of the SOW in the first entry of this issue and confirm your signoff and/or propose any needed updates?  Thanks!

**Comment by jjscheel** [comment]
@haicheng-li, please give some thought as to "sizing" the effort and then extrapolate that to when you think the last PR will be accepted for the work (i.e. we are completely done) with this.  Generally, on-boarding items should take 1-3 months, but that is a guideline, not a rule.

**Comment by atwufei** [comment]
we have posted the initial RFC patch series here:
    https://sourceforge.net/p/valgrind/mailman/message/37850301/

**Comment by jjscheel** [comment]
Thanks, @atwufei!  It is good to see some response in the thread too.

**Comment by jjscheel** [comment]
@haicheng-li, @atwufei, would you kindly provide a brief summary of the information you shared in Tuesday's meeting here?  Thanks!

**Comment by atwufei** [comment]
I have sent out the patches with the prototype of new vector-IR design, Petr gave some positive reviews in general. He will try to verify if this design is also suitable for ARM SVE2, we can move faster after it's proved to be the generic solution for vlen vector.

https://sourceforge.net/p/valgrind/mailman/valgrind-developers/?viewmonth=202307&viewday=18

**Comment by jjscheel** [comment]
Thanks, @atwufei!  It sounds like we are making good progress.

When we have verification that the community is happy with the design and the PR is in good shape, I propose we can declare this item complete.

**Comment by atwufei** [comment]
We have a public tree for this:
    https://github.com/intel/valgrind-rvv/tree/poc-rvv

This branch has the initial code of vector IR, which looks the preferred method for RVV on valgrind. It can run some test cases with RVV instructions.

**Comment by jjscheel** [comment]
@ptomsich, per my email, please review the PR from Wufei and let me know if we can declare the Intel on-boarding as a DevPartner complete.   Thanks!

**Comment by jjscheel** [comment]
@atwufei, I have approval from Philipp for the on-boarding process.  We will now consider Intel a full Development Partner.  Thank you and congratulations.

I'll get updates to the site and ensure we get this announced in the coming week through various RISC-V areas.

**Comment by haicheng-li** [comment]
great to hear the progress, and thanks all for help to make this happen! @jjscheel @atwufei @ptomsich

**Comment by jjscheel** [comment]
@atwufei, any updates?

**Comment by atwufei** [comment]
I've committed a lot more patches to the repo: https://github.com/intel/valgrind-rvv/tree/poc-rvv

The majority of the integer rvv instructions are enabled except a few, it can run coremark compiled with auto-vecterization option, and we expect to complete the integer rvv instructions.  Please note:
1. "enable" means valgrind can run the binary with this instructions, there is still work to make the semantics of memcheck right, although I have tried to make it so, we might not find every memory access problem in rvv instruction, but the bottom line is that there is no side-effect on non-rvv instruction.
2. we have not tested it very much, we try to support more instructions first.
3. no work on fixed-point and floating-point insn yet, they are low priority.

**Comment by jjscheel** [comment]
@atwufei, can you summarize the recent activity here when you have a moment?  Thanks!

**Comment by atwufei** [comment]
After I sent the RVV prototype, Petr, who is the maintainer of valgrind riscv porting, investigated ARM SVE on valgrind and now he has an initial SVE prototype, so it's the time to seek for a unified solution for both and even other vector ISAs.

Regarding to my RVV prototype, Petr's most concern is that "vl" is embedded into every IRType, which he thinks it's too riscv-centric and it has some drawbacks. His proposal is to use "mask" to represent "vl", which I think it's too sve-centric, we need a concrete solution for all the rvv instructions from both function and performance perspectives if we want to go this way. Currently the discussion is still ongoing.

**Comment by jjscheel** [comment]
FYI, an email was received from David Abdurachmanov about build hardware for Valgrind:
> Valgrind upstreaming is stuck
for months now [2]. One of the reasons for that is the lack of
available hardware for development. There are no builders in
http://builder.sourceware.org/ Naturally I looked into cfarm (GCC
Compile Farm) and there aren't many boards either.
>
> [2] https://bugs.kde.org/show_bug.cgi?id=468575

I will work with David to address the issue as best we can.

Please be aware of the RISC-V Developer Board program described here:
https://riscv.org/risc-v-developer-boards/

Interested projects/parties should follow the yellow button to this page and apply for appropriate hardware:
https://riscv.org/risc-v-developer-boards/details/

Please let me know if you hear of any concerns in this area and feel free to make introductions as needed.

**Comment by haicheng-li** [comment]
Fei is working on PoC development of the revised design that removes vl from IR. We will get back to discuss with Petr when PoC is done and tested working fine for ARM side.

**Comment by haicheng-li** [comment]
https://sourceforge.net/p/valgrind/mailman/message/58728269/
the PoC of revised design was done and sent out for review.

**Comment by jjscheel** [comment]
Are we still making progress here?  Anything to share?

BTW, I notice that this is closed.  Any idea when and why that was done?

**Comment by jjscheel** [comment]
Looks like it was closed late last year, likely inadvertently.  Re-opened mostly for tracking purposes.  We can close again if people feel strongly about it, but typically we keep open until PR accepted upstream.

**Comment by haicheng-li** [comment]
Xiao Wang (xiao.w.wang@intel.com) will work on this together with Fei.
his github account is XiaoWang1772

**Comment by jjscheel** [comment]
Thanks, @haicheng-li.  I've added Xiao in Groups.IO and will get an invite out shortly to the 
riscv-devpartner-partners team in the riscv-admin organization.  That will enable access to this and all issues.

**Comment by jjscheel** [comment]
Welcome, @XiaoWang1772!

**Comment by XiaoWang1772** [comment]
I would work with Jun from CAS to continue the effort to support B-ext, V-ext. We need to ramp up on the current project status first. Thanks.

**Comment by haicheng-li** [comment]
Per Xiao, Valgrind riscv base code has been merged in upstream. And Xiao will move to work on other job, so he will quit from this project. We would need some partner to take over the RVV PoC upstreaming work.

**Comment by haicheng-li** [comment]
During Risc-V summit China 2025, I met and f2f talked with Wei Wu @lazyparser about the project background and current status, as well as emailed him the existing PoC patchset & previous design discussions. PLCT will take over this task as there is no other partner wants to work on this project for 3 months.
Thanks @lazyparser & PLCT for the help and effort.

**Comment by lazyparser** [comment]
> During Risc-V summit China 2025, I met and f2f talked with Wei Wu [@lazyparser](https://github.com/lazyparser) about the project background and current status, as well as emailed him the existing PoC patchset & previous design discussions. PLCT will take over this task as there is no other partner wants to work on this project for 3 months. Thanks [@lazyparser](https://github.com/lazyparser) & PLCT for the help and effort.

Thanks! The PLCT Lab is happy to help.

**Comment by Liaoshihua** [comment]
Hello, Jeff @jjscheel, Liao Shihua from PLCT ( @Liaoshihua ) will take over this task and work on this.

**Comment by Liaoshihua** [comment]
In the last month, i ported the Intel's branch to upstream [Liaoshihua/valgrind-riscv](https://github.com/Liaoshihua/valgrind-riscv) and work in implementing zba zbb zbc zbs extension in valgrind. 
But Valgrind riscv port is currently in an unanswered state. The last email from Petr Pavlu, the maintainer of the RISC-V backend, appeared in June 2024. Since then, the RISC-V port has been stagnant.

**Comment by haicheng-li** [comment]
> In the last month, i ported the Intel's branch to upstream [Liaoshihua/valgrind-riscv](https://github.com/Liaoshihua/valgrind-riscv) and work in implementing zba zbb zbc zbs extension in valgrind. But Valgrind riscv port is currently in an unanswered state. The last email from Petr Pavlu, the maintainer of the RISC-V backend, appeared in June 2024. Since then, the RISC-V port has been stagnant.

@Liaoshihua, thanks Shihua for the update. Petr's riscv-port repo should have been obsoleted now as the code has been merged to valgrind upstream repo: https://sourceware.org/git/valgrind.git.  Probably you'd directly work with valgrind upstream to enable the support.

**Comment by Liaoshihua** [comment]
> Probably you'd directly work with valgrind upstream to enable the support.

Yes, my branch point is valgrind_3_26_0 from  https://sourceware.org/git/valgrind.git

**Comment by jjscheel** [comment]
Considering the work complete enough for DevPartners.  Thanks to ALL for your consistent and persistent work here.  You've laid a great foundation the community can decide whether to use.

Closing.
